### DISCUSSION TITLE
Async request aggregation

### DISCUSSION THREAD

**Discussion by artyom-p** [ThreeMammals/Ocelot#1812] [Ideas]
## New Feature

It would be great if Ocelot had an opportunity to aggregate downstream requests asynchronously. Something similar to Rx, or even use Rx for it. It's usefull when you have multiple downstream requests and you want to return the values from downstream requests as you receive them.

For example if there are 3 downstream requests aggregated:

Find()
   - FindInSource1()
   - FindInSource2()
   - FindInSource3()

And as we receive first response from any of  downstream endpoints, send it as part of upstream request response... As far as i know this can be implemented via PushStreamContent, correct me if I'm wrong.

In my own opinion, it's a really nice and usefull feature, especially for user experience, when the frontend is trying to get data trough the gateway.

**Comment by TomPallister** [comment]
@xxsybreedxx yep this would be a good feature! I'm not sure how much effort would be involved but will take a look when I have some free time.

Happy to accept a PR if you fancy having a go yourself!

**Comment by arielmoraes** [comment]
The solution we have for this is to create a configuration that points to the same server (or maybe another) and create the controllers needed for the aggregation. Inside the controller's actions you can then create several tasks and wait all of them and even create new requests that are dependant of any of the previous requests made.

**Now for a possible solution @TomPallister** 
We could have another kind of aggregator that receives a list of tasks after the process of authentication and authorisation, then our code inside the aggregator should be responsible for awaiting the tasks to complete and return the control to the Ocelot pipeline somehow.

**Comment by TomPallister** [comment]
@arielmoraes that in an interesting idea and would be pretty easy to implement I guess.

**Comment by raman-m** [comment]
Ideas, ideas, ideas.... Who will be a developer for these nice ideas? 🤣

**Comment by raman-m** [comment]
> @artyom-p commented [on May 31, 2018](#discussion-5906654)

Hi Artyom!
Are you still with Ocelot?
Where is your [Ocelot fork](https://github.com/artyom-p?tab=repositories)? I don't see it...

**Comment by raman-m** [comment]
> @arielmoraes commented [on Sep 4, 2018](#discussioncomment-7708042)
...and return the control to the Ocelot pipeline **somehow**.

What are the advantages of delegating tasks to another service layer for processing? I'm struggling to grasp your concept. Could you illustrate your idea by submitting a new pull request, please?

**Comment by raman-m** [comment]
@artyom-p 
>And as we receive first response from any of downstream endpoints, send it as part of upstream request response... As far as i know this can be implemented via `PushStreamContent`, correct me if I'm wrong.

Are you sure that `PushStreamContent` is a part of .NET Core? I don't think so.
The `PushStreamContent` class is a part of ASP.NET MVC 4 before .NET Core releases!
These links show that `PushStreamContent` is outdated:
- https://learn.microsoft.com/en-us/search/?terms=PushStreamContent
- https://stackoverflow.com/questions/33892408/pushstreamcontent-in-asp-net-5-mvc-6-is-not-working

So, you proposed something which is not a part of .NET 6+...
But sure, some alternatives should exist...

If you are active contributor, feel free to discuss it here more...
To be fair, I don't see much benefit in making the Aggregator feature completely asynchronous..

**1st, we have to find alternatives in .NET 6+ which we support...
2nd, do you have an intention to contribute as the author of this idea?**
